File: MindCrew-frontend/src/api/apiKey.ts
import request from '@/utils/request'

export interface ApiKey {
  id: number
  name: string
  keyPrefix: string
  allowedKbIds?: string         // JSON string
  monthlyQuota: number
  monthUsed: number
  rateLimitQps: number
  totalCalls: number
  lastUsedAt?: string
  expireAt?: string
  status: 'active' | 'revoked' | 'expired'
  description?: string
  createdBy?: number
  createTime?: string
}

export interface ApiCallLog {
  id: number
  keyId: number
  kbId?: number
  api: string
  question?: string
  statusCode: number
  inputTokens: number
  outputTokens: number
  costCny: number
  latencyMs: number
  ip?: string
  userAgent?: string
  errorMsg?: string
  calledAt: string
}

export interface IssueResult {
  id: number
  rawKey: string          // 完整 key · 仅此一次返回
  prefix: string
  warning: string
}

export const apiKeyApi = {
  /** 生成新 API Key · 返回的 rawKey 必须立即让用户复制保存 */
  issue: (data: {
    name: string
    allowedKbIds: number[]
    monthlyQuota?: number
    rateLimitQps?: number
    expireAt?: string
    description?: string
  }): Promise<any> => request.post('/v2/api-key', data),

  page: (params: { current?: number; size?: number; kbId?: number; status?: string }): Promise<any> =>
    request.get('/v2/api-key/page', { params }),

  /** 11.6 · 列出某 KB 的所有 API Key */
  byKb: (kbId: number): Promise<any> => request.get(`/v2/api-key/by-kb/${kbId}`),

  revoke: (id: number): Promise<any> => request.post(`/v2/api-key/${id}/revoke`),

  updateQuota: (id: number, data: { monthlyQuota?: number; rateLimitQps?: number }): Promise<any> =>
    request.put(`/v2/api-key/${id}/quota`, data),

  updateAllowedKbs: (id: number, allowedKbIds: number[]): Promise<any> =>
    request.put(`/v2/api-key/${id}/kbs`, { allowedKbIds }),

  delete: (id: number): Promise<any> => request.delete(`/v2/api-key/${id}`),

  logs: (params: { current?: number; size?: number; keyId?: number; kbId?: number }): Promise<any> =>
    request.get('/v2/api-key/logs', { params }),
}

File: src/main/java/com/simon/MindCrew/entity/ApiKey.java
package com.simon.MindCrew.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;
import java.time.LocalDateTime;

/**
 * 对外 API Key · 任务 11
 *
 * 字段说明：
 *   key_prefix  · 前缀 mk_xxxxxxxx，用于列表展示（永远不存明文）
 *   key_hash    · 完整 key 的 SHA-256，用于校验（不可逆）
 *   allowed_kb_ids · JSON 数组（11.6 · 一个 key 至少绑 1 个 KB）
 *
 * 安全考虑（铁律 · 务实零 mock）：
 *   - 完整 key 仅在生成那一刻通过 API 返回给前端**一次**
 *   - DB 只存 prefix + SHA-256 hash
 *   - 后端校验时把传入 key 做 SHA-256 后查 hash
 *   - 即便 DB 泄露，攻击者也无法逆推完整 key
 */
@Data
@TableName("api_key")
public class ApiKey {

    @TableId(type = IdType.AUTO)
    private Long id;

    private String name;
    private String keyPrefix;
    private String keyHash;

    /** JSON 数组 · 该 key 可访问的 KB id 列表 */
    private String allowedKbIds;

    private String scopeType;
    private Integer monthlyQuota;
    private Integer rateLimitQps;
    private Integer monthUsed;
    private String monthKey;
    private Long totalCalls;
    private LocalDateTime lastUsedAt;
    private LocalDateTime expireAt;

    /** active / revoked / expired */
    private String status;

    private Long createdBy;
    private String description;

    @TableField(fill = FieldFill.INSERT)
    private LocalDateTime createTime;

    @TableField(fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updateTime;

    @TableLogic
    private Integer deleted;
}
